src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -1340,6 +1340,24 @@ public void SetBadRequestState(BadHttpRequestException ex)
             _keepAlive = false;
         }
 
+        // Normally this would have been rejected, but the developer opted into allowing the bad behavior.
+        public void ReportAllowedBadRequest(BadHttpRequestException ex)
+        {
+            Log.ConnectionBadRequest(ConnectionId, ex);
+            // Set this so DiagnosticSource listeners can observe the exception via the IBadRequestExceptionFeature.
+            // Make sure to unset before exiting the method so the request doesn't actually get rejected.
+            // This means the exception will not be visible to normal middleware.
+            _requestRejectedException = ex;
+
+            const string badRequestEventName = "Microsoft.AspNetCore.Server.Kestrel.BadRequest";
+            if (ServiceContext.DiagnosticSource?.IsEnabled(badRequestEventName) == true)
+            {
+                ServiceContext.DiagnosticSource.Write(badRequestEventName, this);
+            }
+
+            _requestRejectedException = null;
```

Comment:
What happens if an exception is thrown from DiagnosticSource.Write? Do we care that _requestRejectedException isn't null?

tldr should it be a try/finally?

Reply:
I think that exception would cause the request to fail anyways, so _requestRejectedException wouldn't matter.

Reply:
There are different paths that a request goes down on failure when this property is sent. e.g. writing a 400 response